feat(server state): add fetch wrapper to utils and integrate react query into application

ryan-zayne opened the pull request:
## Description

- Added Verbatim option to tsconfig, enforces adding of type directive to type imports (does this automatically if you set vscode codeActions to organize on save, or when you auto import).

- Changed all type imports to use type directive, removed unnecesary React object imports in the process.

- Added Fetch wrapper creator to utils. It creates a fetcher function that extends the barebones fetch api, allowing for things like timeouts, auto promise rejection on network failure, interceptor support and more. 

- Created an instance of the fetcher to be used exclusively with react-query, hence it's baseUrl is the abeg-help-backend

- The instance is also extensible via the config object

- Added React query package, and it's corresponding eslint rules

## Related Issue

- Fixes #123 


## Checklist

- [x] I have reviewed the Contribution Guidelines linked above.
- [x] I have tested my changes thoroughly and ensured that all existing tests pass.
- [x] I have provided clear and concise commit messages.
- [ ] I have updated the project's documentation as necessary.

ryan-zayne committed:
A draft commit....

ryan-zayne committed:
A draft commit....

289Volts commented on lib/utils/fetcher.ts:
```diff
@@ -0,0 +1,14 @@
+const BASE_BACKEND_URL = "http://abeghelp-backend-staging.up.railway.app"; //TODO - Find a way to hide this via environment variables
```
create an env.local file by the time you finalize this PR

ryan-zayne replied:
Speaking of env, do you happen to know how NextJs's bundler handles env variables? 

From what I keep hearing, most bundlers still end up shipping the env variables with the minified code to prod...👀

289Volts replied:
I've not heard about this. I do know that the way to expose env to the client is by using the `NEXT_PUBLIC` prefix. Plus, i will still add the endpoint to vercel manually

ryan-zayne replied:
Fair enough then, I'll leave that worry for later it seems 👍

Danieljs-codes replied:
If the ENV will be used on the client Vercel encrypts it tho

Ryan-Zayne replied:
Fair enough then, will leave this worry for later👍

ryan-zayne replied:
> If the ENV will be used on the client Vercel encrypts it tho

Just what I needed to hear. So it's all safe then👍

On that note tho, which section of the NextJs docs could I find this information? Can't seem to find it just yet

Ryan-Zayne replied:
> > If the ENV will be used on the client Vercel encrypts it tho
> 
> Just what I needed to hear. So it's all safe then👍
> 
> On that note tho, which section of the NextJs docs could I find this information? Can't seem to find it for some reason

ryan-zayne replied:
@Danieljs-codes 👀

1finedev replied:
Just to chime in here, the env encryption done by vercel is env's related to server actions in the app dir, so its not applicable here.

in this case it is perfectly safe to expose the url of our backend endpoints, because you cannot fully hide them anyway, they will be visible in the network tab regardless haha.

It is also best practice to put them in a public ENV variable so it is reusable app wide without copy pasting everywhere

ryan-zayne replied:
> Just to chime in here, the env encryption done by vercel is env's related to server actions in the app dir, so its not applicable here.
> 
> in this case it is perfectly safe to expose the url of our backend endpoints, because you cannot fully hide them anyway, they will be visible in the network tab regardless haha.
> 
> It is also best practice to put them in a public ENV variable so it is reusable app wide without copy pasting everywhere

Thanks for the clarification🤝

ryan-zayne committed:
style(prettier-config): added a two options to prettier, both to reduce styling inconsistencies found.

feat(fetcher): finished with fetch API wrapper, made it configurable and extendable as well

ryan-zayne committed:
style(lint-staged-fixes)

ryan-zayne committed:
refactor: added option for overriding defaults (debatable)

ryan-zayne committed:
Revert "style(lint-staged-fixes)"

This reverts commit 648802b6f4c8635e304812d468a8f5a2c5d6da43.

ryan-zayne committed:
reverts

ryan-zayne committed:
...Draft

ryan-zayne committed:
Corrected typo

babblebey committed:
Merge branch 'dev' into feat(server-state)-add-fetch-wrapper-and-rq

ryan-zayne committed:
Update createFetcherInstance.ts

ryan-zayne committed:
Update createFetcherInstance.ts, with a better base timeout api

ryan-zayne committed:
Fix Eslint error (coding on mobile sucks tbh)

ryan-zayne committed:
Amendment

ryan-zayne committed:
feat: added react-query + devtools + eslint

ryan-zayne committed:
.

ryan-zayne committed:
Revert "."

This reverts commit 15becbc71aabf26355009fcef1058856a8334bf6.

ryan-zayne committed:
updated pkg lock, to stop ci fails

ryan-zayne committed:
feat: done

ryan-zayne committed:
.

Ryan-Zayne commented:
@289Volts @babblebey 

Y'all can come check things out....

And we also have some things to look into as regards to the api usage

ryan-zayne committed:
.

ryan-zayne committed:
refactor: few extra changes for supposedly better error handling

ryan-zayne committed:
refactor: added jsdoc to further document fetchConfig extensions

ryan-zayne committed:
added better error names

ryan-zayne committed:
Added option to disable type truncation

ryan-zayne committed:
.final edit

ryan-zayne committed:
refactor: Patched a few edge cases

refactor: applied principle of least allowance on the fetcher API

ryan-zayne committed:
.

ryan-zayne committed:
.

ryan-zayne committed:
cleaned up implementation

babblebey commented on lib/utils/create-fetcher-instance/create-fetcher-instance.ts:
```diff
@@ -0,0 +1,89 @@
+      // Clean up the timeout
+    } finally {
+      window.clearTimeout(timeoutId);
+    }
```
❤️

I love this thoughts at this point

ryan-zayne replied:
🚀🚀🚀

ryan-zayne committed:
fix

ryan-zayne committed:
.

ryan-zayne committed:
.

babblebey reviewed (approved):
🤲

ryan-zayne committed:
feat: added error interceptor

refactor: patched up and unhandled scenerio in the factory function

ryan-zayne committed:
More cleanups

289Volts commented on components/primitives/Button/button.tsx:
```diff
@@ -1,5 +1,5 @@
-import React, { forwardRef } from "react";
+import { forwardRef } from "react";
 
 interface ButtonProps extends React.ComponentPropsWithRef<"button"> {
```
You deleted React in line 2 and you're still using it here. it didn't cause an error?

ryan-zayne replied:
Yeah, it's the type version that's being used so that doesn't affect anything

289Volts commented on lib/utils/global-type-helpers.ts:
Does this mean we always have to put env values in quotes?

ryan-zayne replied:
Not necessarily, env variables will always be strings whether or not you put them in quotes

ryan-zayne committed:
finallest

1finedev reviewed (approved):
Amazing work ryan, let us keep pushing 

LGTM
